tedana/utils.py

```diff
@@ -49,88 +49,144 @@ def reshape_niimg(data):
-    lthrs = np.squeeze(echo_means[perc_val].T) / 3
-
-    # if multiple samples were extracted per echo, keep the one w/the highest signal
-    if lthrs.ndim > 1:
-        lthrs = lthrs[:, lthrs.sum(axis=0).argmax()]
-
-    # determine samples where absolute value is greater than echo-specific thresholds
-    # and count # of echos that pass criterion
-    masksum = (np.abs(echo_means) > lthrs).sum(axis=-1)
+    assert methods, "No methods provided for adaptive mask generation."
+    assert all([method in ["decay", "dropout", "none"] for method in methods])
+
+    n_samples, n_echos, _ = data.shape
+    adaptive_masks = []
+
+    if "none" in methods:
+        none_adaptive_mask = np.full(n_samples, n_echos, dtype=int)
+        adaptive_masks.append(none_adaptive_mask)
+
+    if ("dropout" in methods) or ("decay" in methods):
+        echo_means = data.mean(axis=-1)  # temporal mean of echos
+
+    if "dropout" in methods:
+        # take temporal mean of echos and extract non-zero values in first echo
+        first_echo = echo_means[echo_means[:, 0] != 0, 0]
+
+        # get 33rd %ile of `first_echo` and find corresponding index
+        # NOTE: percentile is arbitrary
+        # TODO: "interpolation" param changed to "method" in numpy 1.22.0
+        #       confirm method="higher" is the same as interpolation="higher"
+        #       Current minimum version for numpy in tedana is 1.16 where
+        #       there is no "method" parameter. Either wait until we bump
+        #       our minimum numpy version to 1.22 or add a version check
+        #       or try/catch statement.
+        perc = np.percentile(first_echo, 33, interpolation="higher")
+        perc_val = echo_means[:, 0] == perc
+
+        # extract values from all echos at relevant index
+        # NOTE: threshold of 1/3 voxel value is arbitrary
+        lthrs = np.squeeze(echo_means[perc_val].T) / 3
+
+        # if multiple samples were extracted per echo, keep the one w/the highest signal
+        if lthrs.ndim > 1:
+            lthrs = lthrs[:, lthrs.sum(axis=0).argmax()]
+
+        # determine samples where absolute value is greater than echo-specific thresholds
+        # and count # of echos that pass criterion
+        dropout_adaptive_mask = (np.abs(echo_means) > lthrs).sum(axis=-1)
+        adaptive_masks.append(dropout_adaptive_mask)
+
+    if "decay" in methods:
+        # Determine where voxels stop decreasing in signal from echo to echo
+        echo_diffs = np.hstack((np.full((n_samples, 1), -1), np.diff(echo_means, axis=1)))
+        diff_mask = echo_diffs >= 0  # flag where signal is not decreasing
+        last_decreasing_echo = diff_mask.argmax(axis=1)
+        last_decreasing_echo[last_decreasing_echo == 0] = n_echos  # if no increase, set to n_echos
+        adaptive_masks.append(last_decreasing_echo)
+
+    # Retain the most conservative of the selected adaptive mask estimates
+    adaptive_mask = np.minimum.reduce(adaptive_masks)
 
     if mask is None:
```

Comment:
I decided to start with this masking PR, but it might be easier for me to review and have you merge #1060 first. Once #1060 is merged, this block of text would move nearer to before the different types of adaptive masks are calculated.

Reply:
See intro comment. Thinking this over again, it would be nicer to get the updated doc strings and function parameters that are in this PR changed and merged into the other masking PR.